src/components/CheckEditor/checkFormTransformations.ts

```diff
@@ -0,0 +1,339 @@
+  IpVersion,
+  PingSettings,
+  HttpSettings,
+  HttpMethod,
+  HttpSettingsFormValues,
+  Label,
+  TcpSettingsFormValues,
+  TcpSettings,
+  DnsSettingsFormValues,
+  DnsSettings,
+  DNSRRValidator,
+  DnsValidationFormValue,
+  ResponseMatchType,
+  Check,
+} from 'types';
+
+import { CHECK_TYPE_OPTIONS, IP_OPTIONS, DNS_RESPONSE_CODES } from 'components/constants';
+import { checkType } from 'utils';
+
+export function selectableValueFrom<T>(value: T, label?: string): SelectableValue<T> {
+  const labelValue = String(value);
+  return { label: label ?? labelValue, value };
+}
+
+export function fallbackSettings(t: CheckType): Settings {
+  switch (t) {
+    case CheckType.HTTP: {
+      return {
+        http: {
+          method: HttpMethod.GET,
+          ipVersion: IpVersion.V4,
+          noFollowRedirects: false,
+        },
+      };
+    }
+    case CheckType.PING: {
+      return {
+        ping: {
+          ipVersion: IpVersion.V4,
+          dontFragment: false,
+        },
+      };
+    }
+    case CheckType.DNS: {
+      return {
+        dns: {
+          recordType: DnsRecordType.A,
+          server: '8.8.8.8',
+          ipVersion: IpVersion.V4,
+          protocol: DnsProtocol.UDP,
+          port: 53,
+        },
+      };
+    }
+    case CheckType.TCP: {
+      return {
+        tcp: {
+          ipVersion: IpVersion.V4,
+          tls: false,
+        },
+      };
+    }
+    default:
+      throw new Error(`Cannot find values for invalid check type ${t}`);
+  }
+}
+
+const getPingSettingsFormValues = (settings: Settings): PingSettingsFormValues => {
+  const pingSettings = settings.ping ?? (fallbackSettings(CheckType.PING) as PingSettings);
+  return {
+    ...pingSettings,
+    ipVersion: IP_OPTIONS.find(({ value }) => value === settings?.ping?.ipVersion) ?? IP_OPTIONS[1],
+  };
+};
+
+const headersToLabels = (headers: string[] | undefined): Label[] =>
+  headers?.map(header => {
+    const parts = header.split(':', 2);
+    return {
+      name: parts[0],
+      value: parts[1],
+    };
+  }) ?? [];
+
+const getHttpSettingsFormValues = (settings: Settings): HttpSettingsFormValues => {
+  const httpSettings = settings.http ?? (fallbackSettings(CheckType.HTTP) as HttpSettings);
+  return {
+    ...httpSettings,
+    validStatusCodes: httpSettings.validStatusCodes?.map(statusCode => selectableValueFrom(statusCode)) ?? [],
+    validHTTPVersions: httpSettings.validHTTPVersions?.map(httpVersion => selectableValueFrom(httpVersion)) ?? [],
+    method: selectableValueFrom(httpSettings.method),
+    ipVersion: selectableValueFrom(httpSettings.ipVersion),
+    headers: headersToLabels(httpSettings.headers),
+  };
+};
+
+const getTcpSettingsFormValues = (settings: Settings): TcpSettingsFormValues => {
+  const tcpSettings = settings.tcp ?? (fallbackSettings(CheckType.TCP) as TcpSettings);
+  return {
+    ...tcpSettings,
+    ipVersion: selectableValueFrom(tcpSettings.ipVersion),
+  };
+};
+
+interface GetDnsValidationArgs {
```

Comment:
A terser way to define this type that also automatically includes all possible values of `ResponseMatchType`:

`type GetDnsValidationArgs = { [key in ResponseMatchType]?: DNSRRValidator };`

Reply:
Love it! I'll take all the TS tips you got